README


This archive contains the distribution Test-Kwalitee,
version 1.22:

  Test the Kwalitee of a distribution before you release it

This software is copyright (c) 2005 by chromatic.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Test-Kwalitee

1.22      2014-08-16 03:25:59Z
    - do not stop running tests at the first failure (broken since 1.20)

1.21      2014-08-16 00:12:37Z
    - document metrics added in Module::CPANTS::Analyse 0.93_03

1.20      2014-08-10 02:57:00Z (TRIAL RELEASE)
    - running the tests via the import method directly has been deprecated;
      the new kwalitee_ok() function is added to provide a more sane and
      logical layout to your tests.

1.19      2014-07-19 17:01:08Z
    - kwalitee-metrics script now includes the module version as well as the
      name, for each metric provided
    - remove test that is rendered invalid with upcoming changes to
      Test::Builder (1.005+)
    - document new metrics added in Module::CPANTS::Analyse 0.93_01

1.18      2014-01-22 04:47:40Z
    - updated list of available metrics

1.17      2013-10-20 21:49:31Z
    - oops, fixed tests that break when AUTHOR_TESTING is not set

1.16      2013-10-20 21:44:53Z
    - now printing even more diagnostics on error (as much as we have
      available)

1.15      2013-09-25 03:12:20Z
    - re-release with fixed compile test

1.14      2013-09-04 01:38:29Z
    - updated inaccurate test prereq

1.13      2013-08-21 20:03:31Z
    - added missing abstract for kwalitee-metrics script
    - no longer issuing a warning if the test is running from xt/ (see v1.10)

1.12      2013-08-01 23:09:47Z
    - adjusted tests to compensate for changes made in Module::CPANTS::Analyse
      0.88/0.90_01

1.11      2013-07-30 00:27:36Z
    - added missing kwalitee-metrics executable to the distribution

1.10      2013-07-29 00:18:31Z
    - we now issue a warning if running when neither AUTHOR_TESTING or
      RELEASE_TESTING environment variables or set
    - test level adjusted, so location of failing test is correct
    - include a script to dump all metrics ("kwalitee-metrics")

1.09      2013-07-18 17:34:33Z
    - the has_test_pod, has_test_pod_coverage tests have been removed - they
      are classified as 'extra', and have been largely considered to be a bad
      idea anyway (these are often shipped as, and ought to be, in xt/)
    - the extractable test has been removed, as it does nothing in dists
      before there is a tarball present
    - new tests have been added: all standard kwalitee tests that can be run
      on a bare distribution without a tarball

1.08      2013-07-16 00:59:47Z
    - documentation fixed to reflect what indicators are actually available
    - metric names are no longer hardcoded, so Module::CPANTS::Analyse has
      more freedom to add and remove metrics

1.07      2013-06-28 19:22:00Z
    - now the indicators are run in the exact order they are returned
      from Module::CPANTS::Kwalitee::*, as some tests depend on the results of
      earlier tests (charsbar)
    - synopsis updated to recommend a better way to run this module, ensuring
      that it is not run by cpantesters or at installation time
    - we no longer create a function in our namespace for every metric we are
      going to test -- this should not break anyone, as these subs were never
      documented as part of the public API

1.06      2013-05-13 21:15:33Z
    - restore previous behaviour of plan()ing in import, to unbreak some
      dists that didn't follow the docs (which in this case is ok since it's a
      horrible idea for a Test module to plan itself anyway) (v1.05)
    - more diagnostic data is printed when a test fails (RT#85107)

1.05      2013-05-13 00:17:32Z
    - more rigorous testing of output; (in order to make this possible) now we
      do END { done_testing } instead of planning a test count

1.04      2013-04-28 15:55:17Z
    - re-release with bumped developer prereqs to depend on a higher version
      of Test::More for t/00-compile.t and t/00-check-deps.t (RT#84900)

1.03      2013-04-27 12:49:31Z
    - fixed documentation to refer to the proper names of Kwalitee tests
      (RT#24506) (thanks, Gavin Sherlock!)
    - cleaned up partially-botched distribution metadata and README

1.02      2013-03-31 21:58:52Z
    - no operational changes - re-releasing under new management via github
      and Dist::Zilla

1.01      2008-08-14
    - no entry made

1.00      2008-08-14 05:03:02Z
    - fixed code to work with modern Module::CPANTS::Analyse (Elliot Shank,
        Xavier Caron, Heiko Eißfeldt, and probably others)
    - removed old code I don't want to support (Makefile.PL)

0.30      2006-03-03 19:01:56Z
    - fix test plan count for including and excluding tests (Chris Dolan)
    - improve test coverage to check for that, too
    - added signature and signature test

0.20      2006-03-03 07:40:06Z
    - added documentation
    - fixed include and exclude tests (thanks to Chris Dolan)
    - suppress weird CPANTS warnings (they didn't think I would do it this way)

0.10      2006-02-15 05:06:25Z
    - hey, this version actually works
    - releasing to my web site, not the CPAN
    - still needs docs and, ironically, more tests

0.01      2004-04-07 05:57:17Z
    - proof of concept
    - initial release
    - either know what you're doing or don't use this


lib/Test/Kwalitee.pm
use strict;
use warnings;
package Test::Kwalitee;
BEGIN {
  $Test::Kwalitee::AUTHORITY = 'cpan:CHROMATIC';
}
# git description: v1.21-3-ga55ecda
$Test::Kwalitee::VERSION = '1.22';
# ABSTRACT: Test the Kwalitee of a distribution before you release it
# KEYWORDS: testing tests kwalitee CPANTS quality lint errors critic
# vim: set ts=8 sw=4 tw=78 et :

use Cwd;
use Test::Builder 0.88;
use Module::CPANTS::Analyse 0.92;
use namespace::clean;

use parent 'Exporter';
our @EXPORT_OK = qw(kwalitee_ok);

my $Test;
BEGIN { $Test = Test::Builder->new }

sub import
{
    my ($class, @args) = @_;

    # back-compatibility mode!
    if (@args % 2 == 0)
    {
        $Test->level($Test->level + 1);
        my %args = @args;
        my $result = kwalitee_ok(@{$args{tests}});
        $Test->done_testing;
        return $result;
    }

    # otherwise, do what a regular import would do...
    $class->export_to_level(1, @_);
}

sub kwalitee_ok
{
    my (@tests) = @_;

    warn "These tests should not be running unless AUTHOR_TESTING=1 and/or RELEASE_TESTING=1!\n"
        # this setting is internal and for this distribution only - there is
        # no reason for you to need to circumvent this check in any other context.
        # Please DO NOT enable this test to run for users, as it can fail
        # unexpectedly as parts of the toolchain changes!
        unless $ENV{_KWALITEE_NO_WARN} or $ENV{AUTHOR_TESTING} or $ENV{RELEASE_TESTING}
            or (caller)[1] =~ /^xt/
            or ((caller)[0]->isa(__PACKAGE__) and (caller(1))[1] =~ /^xt/);

    my @run_tests = grep { /^[^-]/ } @tests;
    my @skip_tests = map { s/^-//; $_ } grep { /^-/ } @tests;

    # These don't really work unless you have a tarball, so skip them
    push @skip_tests, qw(extractable extracts_nicely no_generated_files
        has_proper_version has_version manifest_matches_dist);

    # MCA has a patch to add 'needs_tarball', 'no_build' as flags
    my @skip_flags = qw(is_extra is_experimental needs_db);

    my $basedir = cwd;

    my $analyzer = Module::CPANTS::Analyse->new({
        distdir => $basedir,
        dist    => $basedir,
        # for debugging..
        opts => { no_capture => 1 },
    });

    my $ok = 1;

    for my $generator (@{ $analyzer->mck->generators })
    {
        $generator->analyse($analyzer);

        for my $indicator (sort { $a->{name} cmp $b->{name} } @{ $generator->kwalitee_indicators })
        {
            next if grep { $indicator->{$_} } @skip_flags;

            next if @run_tests and not grep { $indicator->{name} eq $_ } @run_tests;

            next if grep { $indicator->{name} eq $_ } @skip_tests;

            my $result = _run_indicator($analyzer->d, $indicator);
            $ok &&= $result;
        }
    }

    return $ok;
}

sub _run_indicator
{
    my ($dist, $metric) = @_;

    my $subname = $metric->{name};
    my $ok = 1;

    $Test->level($Test->level + 1);
    if (not $Test->ok( $metric->{code}->($dist), $subname))
    {
        $ok = 0;
        $Test->diag('Error: ', $metric->{error});

        # NOTE: this is poking into the analyse structures; we really should
        # have a formal API for accessing this.

        # attempt to print all the extra information we have
        my @details;
        push @details, $metric->{details}->($dist)
            if $metric->{details} and ref $metric->{details} eq 'CODE';
        push @details,
            (ref $dist->{error}{$subname}
                ? @{$dist->{error}{$subname}}
                : $dist->{error}{$subname})
            if defined $dist->{error} and defined $dist->{error}{$subname};
        $Test->diag("Details:\n", join("\n", @details)) if @details;

        $Test->diag('Remedy: ', $metric->{remedy});
    }
    $Test->level($Test->level - 1);

    return $ok;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Kwalitee - Test the Kwalitee of a distribution before you release it

=head1 VERSION

version 1.22

=head1 SYNOPSIS

In a separate test file:

  use Test::More;
  BEGIN {
      plan skip_all => 'these tests are for release candidate testing'
          unless $ENV{RELEASE_TESTING};
  }

  use Test::Kwalitee 'kwalitee_ok';
  kwalitee_ok();
  done_testing;

=head1 DESCRIPTION

=for stopwords CPANTS

Kwalitee is an automatically-measurable gauge of how good your software is.
That's very different from quality, which a computer really can't measure in a
general sense.  (If you can, you've solved a hard problem in computer science.)

In the world of the CPAN, the CPANTS project (CPAN Testing Service; also a
funny acronym on its own) measures Kwalitee with several metrics.  If you plan
to release a distribution to the CPAN -- or even within your own organization
-- testing its Kwalitee before creating a release can help you improve your
quality as well.

C<Test::Kwalitee> and a short test file will do this for you automatically.

=head1 USAGE

Create a test file as shown in the synopsis.  Run it.  It will run all of the
potential Kwalitee tests on the current distribution, if possible.  If any
fail, it will report those as regular diagnostics.

If you ship this test, it will not run for anyone else, because of the
C<RELEASE_TESTING> guard. (You can omit this guard if you move the test to
xt/release/, which is not run automatically by other users.)

=head1 FUNCTIONS

=head2 kwalitee_ok

With no arguments, runs all standard metrics.

To run only a handful of tests, pass their name(s) to the C<kwalitee_ok>
function:

  kwalitee_ok(qw( use_strict has_tests ));

To disable a test, pass its name with a leading minus (C<->):

  kwalitee_ok(qw( -use_strict has_readme ));

=head1 BACK-COMPATIBILITY MODE

Previous versions of this module ran tests directly via the C<import> sub, like so:

    use Test::Kwalitee;
    # and that's it!

...but this is problematic if you need to perform some setup first, as you
would need to do that in a C<BEGIN> block, or manually call C<import>. This is
messy!

However, this calling path is still available, e.g.:

  use Test::Kwalitee tests => [ qw( use_strict has_tests ) ];

=head1 METRICS

The list of each available metric currently available on your
system can be obtained with the C<kwalitee-metrics> command (with
descriptions, if you pass C<--verbose> or C<-v>, but
as of Test::Kwalitee 1.19 and L<Module::CPANTS::Analyse> 0.93_01, the tests include:

=over 4

=item *

has_abstract_in_pod

Does the main module have a C<=head1 NAME> section with a short description of the distribution?

=item *

has_buildtool

Does the distribution have a build tool file?

=for stopwords changelog

=item *

has_changelog

Does the distribution have a changelog?

=item *

has_humanreadable_license

Is there a C<LICENSE> section in documentation, and/or a F<LICENSE> file
present?

=item *

has_license_in_source_file

Is there license information in any of the source files?

=item *

has_manifest

Does the distribution have a F<MANIFEST>?

=item *

has_meta_yml

Does the distribution have a F<META.yml> file?

=item *

has_readme

Does the distribution have a F<README> file?

=item *

has_tests

Does the distribution have tests?

=item *

meta_yml_conforms_to_known_spec

=for stopwords recognised

Does META.yml conform to any recognised META.yml specification?
(For specs see
L<http://module-build.sourceforge.net/META-spec-current.html>)

=item *

meta_yml_is_parsable

Can the F<META.yml> be parsed?

=item *

meta_json_conforms_to_known_spec

Does META.json conform to the recognised META.json specification?
(For specs see L<CPAN::Meta::Spec>)

=item *

meta_json_is_parsable

Can the F<META.json> be parsed?

=item *

no_broken_auto_install

Is the distribution using an old version of L<Module::Install>? Versions of
L<Module::Install> prior to 0.89 do not detect correctly that C<CPAN>/C<CPANPLUS>
shell is used.

=item *

no_broken_module_install

Does the distribution use an obsolete version of L<Module::Install>?
Versions of L<Module::Install> prior to 0.61 might not work on some systems at
all. Additionally if the F<Makefile.PL> uses the C<auto_install()>
feature, you need at least version 0.64. Also, 1.04 is known to be broken.

=item *

no_symlinks

Does the distribution have no symlinks?

=item *

use_strict

Does the distribution files all use strict?

=back

=head1 ACKNOWLEDGEMENTS

=for stopwords Klausner Dolan

With thanks to CPANTS and Thomas Klausner, as well as test tester Chris Dolan.

=head1 SEE ALSO

=over 4

=item *

L<kwalitee-metrics> (in this distribution)

=item *

L<Module::CPANTS::Analyse>

=item *

L<App::CPANTS::Lint>

=item *

L<Test::Kwalitee::Extra>

=item *

L<Dist::Zilla::Plugin::Test::Kwalitee>

=item *

L<Dist::Zilla::Plugin::Test::Kwalitee::Extra>

=item *

L<Dist::Zilla::App::Command::kwalitee>

=back

=head1 AUTHORS

=over 4

=item *

chromatic <chromatic@wgz.org>

=item *

Karen Etheridge <ether@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by chromatic.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 CONTRIBUTORS

=for stopwords Gavin Sherlock David Steinbrunner Nathan Haigh Zoffix Znet Kenichi Ishigaki

=over 4

=item *

Gavin Sherlock <sherlock@cpan.org>

=item *

David Steinbrunner <dsteinbrunner@pobox.com>

=item *

Nathan Haigh <nathanhaigh@ukonline.co.uk>

=item *

Zoffix Znet <cpan@zoffix.com>

=item *

Kenichi Ishigaki <ishigaki@cpan.org>

=back

=cut


lib/Test/Kwalitee/Conflicts.pm
package # hide from PAUSE
    Test::Kwalitee::Conflicts;

use strict;
use warnings;

use Dist::CheckConflicts
    -dist      => 'Test::Kwalitee',
    -conflicts => {
        'Dist::Zilla::Plugin::Test::Kwalitee' => '2.04',
    },

;

1;

# ABSTRACT: Provide information on conflicts for Test::Kwalitee

__END__

=pod

=encoding UTF-8

=for :stopwords chromatic Karen Etheridge David Steinbrunner Gavin Sherlock Kenichi
Ishigaki Nathan Haigh Zoffix Znet

=head1 NAME

Test::Kwalitee::Conflicts - Provide information on conflicts for Test::Kwalitee

=head1 VERSION

version 1.18

=head1 AUTHORS

=over 4

=item *

chromatic <chromatic@wgz.org>

=item *

Karen Etheridge <ether@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by chromatic.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL
# This Build.PL for Test-Kwalitee was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.005
if (eval 'use Module::Build::Tiny 0.037; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for Test-Kwalitee was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.006.
    use 5.006;
    # use Module::Build::Tiny 0.037;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.037"
      },
      "dist_abstract" => "Test the Kwalitee of a distribution before you release it",
      "dist_author" => [
        "chromatic <chromatic\@wgz.org>",
        "Karen Etheridge <ether\@cpan.org>"
      ],
      "dist_name" => "Test-Kwalitee",
      "dist_version" => "1.22",
      "license" => "perl",
      "module_name" => "Test::Kwalitee",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "Cwd" => 0,
        "Exporter" => 0,
        "Module::CPANTS::Analyse" => "0.92",
        "Test::Builder" => "0.88",
        "namespace::clean" => 0,
        "parent" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [
        "script/kwalitee-metrics"
      ],
      "test_requires" => {
        "CPAN::Meta::Check" => "0.007",
        "CPAN::Meta::Requirements" => 0,
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Test::Deep" => 0,
        "Test::More" => "0.88",
        "Test::Tester" => "0.108",
        "Test::Warnings" => "0.009",
        "if" => 0
      }
    );


    my %fallback_build_requires = (
      "CPAN::Meta::Check" => "0.007",
      "CPAN::Meta::Requirements" => 0,
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Test::Deep" => 0,
      "Test::More" => "0.88",
      "Test::Tester" => "0.108",
      "Test::Warnings" => "0.009",
      "if" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}


Makefile.PL
# This Makefile.PL for Test-Kwalitee was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.013
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.24.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.037',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Test the Kwalitee of a distribution before you release it",
  "AUTHOR" => "chromatic <chromatic\@wgz.org>, Karen Etheridge <ether\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.037"
  },
  "DISTNAME" => "Test-Kwalitee",
  "EXE_FILES" => [
    "script/kwalitee-metrics"
  ],
  "LICENSE" => "perl",
  "NAME" => "Test::Kwalitee",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "Exporter" => 0,
    "Module::CPANTS::Analyse" => "0.92",
    "Test::Builder" => "0.88",
    "namespace::clean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "CPAN::Meta::Check" => "0.007",
    "CPAN::Meta::Requirements" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "Test::Tester" => "0.108",
    "Test::Warnings" => "0.009",
    "if" => 0
  },
  "VERSION" => "1.22",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta::Check" => "0.007",
  "CPAN::Meta::Requirements" => 0,
  "Cwd" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::CPANTS::Analyse" => "0.92",
  "Test::Builder" => "0.88",
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "Test::Tester" => "0.108",
  "Test::Warnings" => "0.009",
  "if" => 0,
  "namespace::clean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



